src/lib/base.jl

```diff
@@ -57,7 +57,8 @@ grad_mut(ch::Channel) = Channel(ch.sz_max)
 
 @adjoint! function put!(ch::Channel, x)
   put!(ch, x), function (ȳ)
-    x̄ = take!(grad_mut(__context__, ch))
+    x̄ = grad_mut(__context__, ch)
+    x̄ = isready(x̄) ? take!(x̄) : x̄
```

Comment:
This need a comment.
I don't understand why we would do this.
`x̄ ` is a`Channel`
`take!` on a `Channel` blocks if not ready.

Are we just doing it early if it is already ready?
Don't was just absolutely have to do this before `accum` completes?
So `accum` must  do blocking `take!`?


(Also maybe not great to reassign a local variable with a different type. Mostly the optimizer can fix it, but somehting it results in being unstable. More importantly it is harder to read. I just never reassign variables :grin:)

Reply:
Reassigning variables is taken.

The issue is that we were taking from a consumed channel at one point.

Reply:
Shouldn't that error?

Reply:
It blocks iirc, since it may have something put back into it at a later time.

Reply:
`accum(::Channel, ȳ)`  isn't defined, is it?
If we don't take from it we can't do the next line

Reply:
It falls back to `accum(<:Any, ::Nothing)` and moves on from there. The change in behaviour of using a channel for syncing was from 1.5. I wanted to use the `taskref` kwarg but for `@async` we don't control how that's constructed, so it's a bit awkward to get the references back.

If a channel is closed it shouldn't have a "partial" for it anyway.

Reply:
I am still not following.
what makes it certain that `ȳ===nothing` just because it isn't ready?
If that is the case why do we check if we can take? why not just check if `ȳ===nothing` ?

Reply:
Missed this completely. Is there an `isclosed` on Channels? Further, there is a case for whether it can be taken from later, so its awkward to decide when to wait for a Channel to be populated vs when it will block endlessly.

We aren't certain that `ȳ===nothing`, we can add a check here. Gradients for Channels themselves would be a `nothing`/ `DoesNotExist` right? The elements of the Channels themselves can have the expected gradients, but the Channel itself is needed only for the references to the elements. The problem is that if we `take!` here and the channel isn't populated we will block.

Reply:
> Missed this completely. Is there an isclosed on Channels?

There is `isopen`

>  Gradients for Channels themselves would be a nothing/ DoesNotExist right? 

I don't know, My instincts say that  `Channels` are basically the same as `Vector`, and so the gradient of a channel is another channel with the gradients of all the elements.
But it seems we don't do that.
So I don't really get this code, i clearly do not have enough of picture of what is going on.

Reply:
Cool, will update it accordingly.

I agree that that makes sense.  We can have it capture all the pullbacks in a different channel which we can take from if it is populated.